### PR TITLE
Test connection with op-program

### PR DESCRIPTION
**Pull request by rbonichon**
This commit contains:
- 1 unit test for bidirectional channels
- 1 integration test (a shell script and a part of our optimism VM)
- fixes to the issues found during the testing phase

### CODE REVIEW

File: optimism/src/main.rs [hunk 1]
```diff
@@ -97,23 +109,26 @@ fn cli() -> VmConfiguration {
...
+            host,
+        },
+        test_preimage_read.map(|x| x.to_string()),
+    )
 }
 
-pub fn main() -> ExitCode {
-    let configuration = cli();
-
+pub fn nominal(configuration: VmConfiguration) -> ExitCode {
```
Comment (dannywillems) [inline optimism/src/main.rs:131]: Curious: what does it mean "nominal" in this case?

File: optimism/test_preimage_read.sh [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+(cd ${OP_DIR} && make op-program)
+
+# LOCATION of cast binary
+# It is to be found in foundry 
+# https://github.com/foundry-rs/foundry
+CAST="/Users/richard/foundry-nightly-5be158ba6dc7c798a6f032026fe60fc01686b33b/target/debug/cast"
+
+# Defaults to dw's servers
+L1RPC=${1:-"http://65.108.142.23:19545"}
+L2RPC=${2:-"http://65.108.142.23:28545"}
```
Comment (dannywillems) [inline optimism/test_preimage_read.sh:17]: Same

File: optimism/test_preimage_read.sh [hunk 3]
```diff
@@ -0,0 +1,87 @@
+#!/bin/sh
+# Test preimage read from VM request to server response
+
+# Location of OPTIMISM source
+OP_DIR=~/zkvm/myoptimism
```
Comment (dannywillems) [inline optimism/test_preimage_read.sh:5]: Should be an argument of the script.

Comment (dannywillems) [reply optimism/test_preimage_read.sh:5]: Or the submodule `ethereum-optimism`

File: optimism/test_preimage_read.sh [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+L1_FINALIZED_NUMBER=$($CAST block finalized --rpc-url "${L1RPC}" -f number)
+L1_FINALIZED_HASH=$($CAST block "${L1_FINALIZED_NUMBER}" --rpc-url "${L1RPC}" -f hash)
+
+OUTPUT_INDEX=$($CAST call --rpc-url "${L1RPC}" "${L2_OUTPUT_ORACLE}" 'getL2OutputIndexAfter(uint256) returns(uint256)' "${L2_FINALIZED_NUMBER}")
+OUTPUT_INDEX=$((OUTPUT_INDEX-1))
+
+OUTPUT=$($CAST call --rpc-url "${L1RPC}" "${L2_OUTPUT_ORACLE}" 'getL2Output(uint256) returns(bytes32,uint128,uint128)' "${OUTPUT_INDEX}")
+OUTPUT_ROOT=$(echo ${OUTPUT} | cut -d' ' -f 1)
+OUTPUT_TIMESTAMP=$(echo ${OUTPUT} | cut -d' ' -f 2)
```
Comment (dannywillems) [inline optimism/test_preimage_read.sh:33]: Unused

File: optimism/test_preimage_read.sh [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+# Test preimage read from VM request to server response
+
+# Location of OPTIMISM source
+OP_DIR=~/zkvm/myoptimism
+
+# Rebuild op-program
+(cd ${OP_DIR} && make op-program)
+
+# LOCATION of cast binary
+# It is to be found in foundry
```
Comment (dannywillems) [inline optimism/test_preimage_read.sh:11]:
```suggestion
# It is to be found in foundry
```

File: optimism/test_preimage_read.sh [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+
+
+NETWORK=sepolia
+
+
+RUST_LOG=debug cargo run -r --bin kimchi_optimism -- --preimage-db-dir ${DATADIR}  -- \
+    ${OP_DIR}/op-program/bin/op-program \
+    --log.level DEBUG \
+    --l1 ${L1RPC} \
```
Comment (dannywillems) [inline optimism/test_preimage_read.sh:76]: Nit (and not only for this PR): we should use double-quotes. As we do have some (minor) bash scripts, we should start using a shellcheck.

File: optimism/test_preimage_read.sh [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+# Rebuild op-program
+(cd ${OP_DIR} && make op-program)
+
+# LOCATION of cast binary
+# It is to be found in foundry 
+# https://github.com/foundry-rs/foundry
+CAST="/Users/richard/foundry-nightly-5be158ba6dc7c798a6f032026fe60fc01686b33b/target/debug/cast"
+
+# Defaults to dw's servers
+L1RPC=${1:-"http://65.108.142.23:19545"}
```
[the remark this answers is not in the document]
Comment (dannywillems) [reply optimism/test_preimage_read.sh:16]: You can source the file `rpcs.sh`.

File: optimism/test_preimage_read.sh (hunk 6 above)
Comment (dannywillems) [reply optimism/test_preimage_read.sh:76]: Open https://github.com/o1-labs/proof-systems/issues/1473

File: optimism/src/main.rs [hunk 8]
```diff
@@ -152,3 +167,71 @@ pub fn main() -> ExitCode {
...
+        .collect::<Result<Vec<_>, io::Error>>()
+        .unwrap();
+
+    // Just take 10 elements at random to test
+    let how_many = 10_usize;
+    let max = paths.len();
+    let mut rng = rand::thread_rng();
+
+    let mut selected: Vec<PathBuf> = vec![PathBuf::new(); how_many];
+    for pb in selected.iter_mut().take(how_many) {
```
Comment (dannywillems) [inline optimism/src/main.rs:193]: Nit: `take(how_many)` is useless no?

**Review by dannywillems — CHANGES_REQUESTED** [review]
LGTM. The test script works. Nice seeing it in action! Good job!
I request changes regarding my previous comments. I will accept after. I will ask you questions off-review :).

**Comment by dannywillems** [comment]
I don't know if it is expected, but we get:
```
t=2023-12-07T19:34:31+0100 lvl=eror msg="failed to close pre-image server" err="signal: interrupt"
```
when stopping the zkVM with `CTRL-C`.

Not urgent IMHO, and definitely not in the scope of this PR, but would be nice to open a follow-up.

File: optimism/src/main.rs [hunk 9]
```diff
@@ -152,3 +167,71 @@ pub fn main() -> ExitCode {
...
+        debug!("Asking for preimage");
+        let preimage = po.get_preimage(key);
+        let got = hex::encode(preimage.get()).to_string();
+
+        assert_eq!(expected, got);
+    }
+    ExitCode::SUCCESS
+}
+
+pub fn main() -> ExitCode {
```
Comment (dannywillems) [inline optimism/src/main.rs:226]: For a future improvement (open follow-up): I would make explicit the split between test and real environment.

File: optimism/test_preimage_read.sh [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+L2_HEAD_NUMBER=$(echo ${OUTPUT} | cut -d' ' -f 3)
+L2_HEAD=$($CAST block "${L2_HEAD_NUMBER}" --rpc-url "${L2RPC}" -f hash)
+
+set -x
+
+# Location of prepopulated OP DB
+# Tune as needed
+DATADIR=${OP_DIR}/op_program.db/
+
+# If you actually need to populate the OP DB, run
```
[the remark this answers is not in the document]
Comment (rbonichon) [reply optimism/test_preimage_read.sh]: done

File: optimism/test_preimage_read.sh (hunk 3 above)
Comment (rbonichon) [reply optimism/test_preimage_read.sh]: Removed

File: optimism/test_preimage_read.sh (hunk 2 above)
Comment (rbonichon) [reply optimism/test_preimage_read.sh]: done

File: optimism/test_preimage_read.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (rbonichon) [reply optimism/test_preimage_read.sh]: done

File: optimism/test_preimage_read.sh (hunk 5 above)
Comment (rbonichon) [reply optimism/test_preimage_read.sh]: Done

File: optimism/src/main.rs (hunk 8 above)
Comment (rbonichon) [reply optimism/src/main.rs]: Yeah

File: optimism/src/main.rs (hunk 9 above)
Comment (rbonichon) [reply optimism/src/main.rs:226]: Indeed. Yeah sure.

Comment (rbonichon) [reply optimism/src/main.rs:226]: Opened https://github.com/o1-labs/proof-systems/issues/1486

**Comment by rbonichon** [comment]
> I don't know if it is expected, but we get:
> 
> ```
> t=2023-12-07T19:34:31+0100 lvl=eror msg="failed to close pre-image server" err="signal: interrupt"
> ```
> 
> when stopping the zkVM with `CTRL-C`.
> 
> Not urgent IMHO, and definitely not in the scope of this PR, but would be nice to open a follow-up.

Opened https://github.com/o1-labs/proof-systems/issues/1487

**Pull request merged**